3891: Check for validity of segment start/end time during segment generation

siddharthteotia opened the pull request:
This will allow us to fail-fast during segment generation if there is incorrect data as opposed to detecting this later during segment upload. The check during segment upload is still retained though

siddharthteotia commented:
cc @mcvsubbu, @jackjlli

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -390,6 +397,43 @@ void writeMetadata()
+      endMillis = ((Integer) endTime).longValue();
+    } else if (startTime instanceof String && endTime instanceof String) {
+      startMillis = Long.parseLong((String)startTime);
+      endMillis = Long.parseLong((String)endTime);
```
I am not quite sure about the type comparison. Have some questions on the format of time that is incoming here at the time of generation v/s how it is stored. By looking at the code in SegmentMetadataImpl, seems like it is stored as string and is always parsed as long

siddharthteotia commented:
@jackjlli , @mcvsubbu , I have been debugging the 31 unit test failures as reported here:
https://travis-ci.org/apache/incubator-pinot/jobs/550998101

Some of them I have fixed like MultiplePinotSegmentRecordReaderTest.  This test was generating its own test data and I tweaked it to account for generating time column values within acceptable range. Similarly, the unit test I added to test the newly added in SegmentGenerationWithTimeColumnTest  also piggy-backed on the random data generation in the test code. These tests first put the dummy data in a list of GenericRow and create a record reader with the schema and generic row set. Later SegmentIndexCreationDriver is initialized with the record reader for dummy data

However, many others like IntArraysTest, ColumnMetadataTest, SegmentMetadataImplTest, SegmentPrepropcessorTest seem to be reading the test data from avro file and thus there is no control in the test code over the data. SegmentIndexCreationDriver is initialized just with the generation config that internally has details of path to avro test dataset.

I am wondering how to fix these kinds of tests now that segment generation code does the time column validity check and the input data is in avro file?

I am still going through rest of the test failures but wanted an opinion w.r.t fixing tests that use external dataset.

jackjlli commented:
@siddharthteotia can we try modifying the test data? There's no point to allow generating un-pushable data. Thanks!

siddharthteotia commented:
@jackjlli , for the tests that I have fixed and the ones that I added, test data was generated by the test code using java random. However, for the other set of tests, the data is not generated by test code itself. Instead, it is an external avro file and we use that to build a segment. So no code change seems possible to change the data. As discussed offline, I will explore using avro-tools and see if the data in time column data in avro file can be updated.

mayankshriv commented:
@siddharthteotia One work-around would be write some throw-away java code that reads in the avro in question and writes another avro with valid/allowed time stamps. Should be pretty straight forward to do that.

siddharthteotia commented:
> @siddharthteotia One work-around would be write some throw-away java code that reads in the avro in question and writes another avro with valid/allowed time stamps. Should be pretty straight forward to do that.

@mayankshriv , thanks for the suggestion. Yes that is what I was doing (after intially trying to put the avro file in a hive table and then updating it via sql). After a brief discussion with @mcvsubbu , we also decided that it will be good to have an option in SegmentGeneratorConfig to enable/disable (if at all needed) this time validity check in segment generation code path. The check is enabled by default and we can see later if realtime code has any surprises in production -- in that case we can disable it. 

Secondly, we also discussed about the changes to failing unit tests. Very few of them were fixed by changing the data generation code. For rest of them, I was in the middle of changing the avro datasets. However, for current PR, I have just explicitly disabled the check for these unit tests that use avro file. I have another patch in progress that cleans up the avro data through a simple java program that updates the records. 

Also, a unit test has been added to test the exception code in SegmentGenerationWithTimeColumnTest.java

siddharthteotia commented:
@jackjlli . @mayankshriv , @mcvsubbu 

Got a clean build. Please review the changes when you get a chance. It took some iterations to make the time check code robust since it has to work with multiple TimeSpecs and both EPOCH and SIMPLE_DATE_FORMAT time column types.

Coming to tests where validity check is explicitly disabled -- as I mentioned earlier, I will create a PR (once this is merged) to clean up the avro data sets used by these tests. 

Some unit tests have been changed to generate valid data (for e.g PinotSegmentUtil) and a new unit test has been added to SegmentGenerationTimecolumnTest to test the newly added exception code in SegmentColumnarIndexCreator

I am seeing if a couple of more unit tests can be added. Meanwhile, it would be great to get feedback on the changes.

jackjlli commented on pinot-core/src/test/java/org/apache/pinot/core/segment/index/creator/SegmentGenerationWithTimeColumnTest.java:
```diff
@@ -81,12 +84,25 @@ public void testSimpleDateSegmentGeneration()
+    boolean error = false;
+    try {
+      Schema schema = createSchema(false);
+      buildSegment(schema, false, true);
```
You can add a statement below this line like:
```
Assert.fail("...");
```
so that it will guarantee to fail the test if `buildSegment()` passes.

siddharthteotia replied:
Yes, we could do that. However, it won't let me test for the existence of message - exception detail. Right?

jackjlli replied:
I think it's ok not to have assertion on the content of the exception; if the exception message changes in the future we'll still need to modify all the checks here. You can just write the code like this:
```
try {
  blabla();
  Assert.fail("...");
} catch (Exception e) {
  // Expected.
}
```

siddharthteotia replied:
Done.

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -391,6 +399,102 @@ void writeMetadata()
+    if (!TimeUtils.timeValueInValidRange(start) || !TimeUtils.timeValueInValidRange(end)) {
+      final Date minDate = new Date(TimeUtils.getValidMinTimeMillis());
+      final Date maxDate = new Date(TimeUtils.getValidMaxTimeMillis());
+      LOGGER.error(
```
Done

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/core/data/readers/PinotSegmentUtil.java:
```diff
@@ -47,16 +52,16 @@ private PinotSegmentUtil() {
   public static List<GenericRow> createTestData(Schema schema, int numRows) {
     List<GenericRow> rows = new ArrayList<>();
-    final Random random = new Random();
+    final ThreadLocalRandom random = ThreadLocalRandom.current();
```
For this class, there is no other use except for the fact that ThreadLocalRandom has the method nextLong(start, end) whereas java.util.Random doesn't provide any such method to randomly generate long values between a given lower and upper bound

siddharthteotia commented:
Addressed the latest review comment. The build is clean

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -391,6 +399,134 @@ void writeMetadata()
+    long start;
+    long end;
+
+    if (startTime instanceof Long && endTime instanceof Long) {
```
I'm thinking whether we can firstly check `startTime` ad `endTime` are of the same class, then use switch block to differentiate the cases on either one of the objects. By doing so we don't need to keep a list of `if else block` here. But you can decide.

siddharthteotia replied:
Done

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -391,6 +399,134 @@ void writeMetadata()
+    } else {
+      final StringBuilder err = new StringBuilder();
+      err.append("Unable to interpret type of time column value. Failed to validate start and end time of segment")
+          .append(" uninterpreted type: ")
```
Could you re-format this file with Pinot check style?

siddharthteotia replied:
Done

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java:
```diff
@@ -391,6 +399,134 @@ void writeMetadata()
+   * @param endTime segment end time
+   * @param segmentName segment name
+   */
+  private void checkTime(final SegmentGeneratorConfig config, final Object startTime, final Object endTime,
```
we already have a time interval checker that the controller uses in the segment push code path. Can we re-use that (if necessary, move it to some common place)

siddharthteotia replied:
The method in SegmentValidator does the check using segment metadata and compares the startMillis and endMillis against the valid range. 

I have written a common utility method in TimeUtils that takes the start and end time as millis from epoch and compared them against the valid range. This ensures that if in future the validity comparison logic changes, we can do it in single place and it will impact both scenarios -- check during generation and check during segment upload in controller.

It didn't seem feasible to abstract the complete method into a common method since new code (in segment generation), handles a lot more information -- data types (INT, LONG, STRING) along with time units (DAYS, MILLIS, HOURs etc) and whether we have EPOCH or SIMPLE_DATE_FORMAT. This also means that if the check fails, I dump few additional information into logger and error message. 

Let me know if this is fine.

siddharthteotia commented:
> also, we should set the check to false in the realtime segment generator code path

Done

siddharthteotia commented:
Addressed latest review comments

siddharthteotia commented:
The build is clean and all latest review comments have been addressed. Please take a look when you get a chance.

mcvsubbu commented on pinot-core/src/main/java/org/apache/pinot/core/realtime/converter/RealtimeSegmentConverter.java:
```diff
@@ -95,6 +95,7 @@ public void build(@Nullable SegmentVersion segmentVersion, ServerMetrics serverM
       reader = new RealtimeSegmentRecordReader(realtimeSegmentImpl, dataSchema, sortedColumn);
     }
     SegmentGeneratorConfig genConfig = new SegmentGeneratorConfig(dataSchema);
+    genConfig.setCheckTimeColumnValidityDuringGeneration(false);
```
Good to add a comment here saying that we don't want realitme consumption to stop (if the exception fires, consumption _will_ stop) in case application is pushing bad data, so we explicitly ignore the time column check while generating realtime segments.

siddharthteotia replied:
Done.

mcvsubbu commented on pinot-core/src/test/java/org/apache/pinot/core/plan/maker/MetadataAndDictionaryAggregationPlanMakerTest.java:
```diff
@@ -94,6 +94,7 @@ public void buildSegment()
     segmentGeneratorConfig.setTableName("testTable");
     segmentGeneratorConfig.setSegmentName(SEGMENT_NAME);
     segmentGeneratorConfig.setOutDir(INDEX_DIR.getAbsolutePath());
+    segmentGeneratorConfig.setCheckTimeColumnValidityDuringGeneration(false);
```
maybe add a comment in these places as to the reason we are not checking for time column validity. I believe the reason is that the input avro files have invalid values for time column.

siddharthteotia replied:
Good point. Done

siddharthteotia commented:
Rebased

siddharthteotia commented:
It looks like the build (for integration tests) was stalled. The log seems to indicate it was running HadoopSegmentBuildPushOfflineClusterIntegrationTest at that point. The test passes locally. May be we can retrigger?
